src/Mod/TechDraw/Gui/SingleInsertTool.cpp

```diff
@@ -0,0 +1,215 @@
+            || obj->isDerivedFrom<App::Link>()) {
+            is_linked = true;
+        }
+        // If parent of the obj is a link to another document, we possibly need to treat non-link obj as linked, too
+        // 1st, is obj in another document?
+        if (obj->getDocument() != cmd.getDocument()) {
+            std::set<App::DocumentObject*> parents = obj->getInListEx(true);
+            for (auto& parent : parents) {
+                // Only consider parents in the current document, i.e. possible links in this View's document
+                if (parent->getDocument() != cmd.getDocument()) {
+                    continue;
+                }
+                // 2nd, do we really have a link to obj?
+                if (parent->isDerivedFrom<App::LinkElement>()
+                    || parent->isDerivedFrom<App::LinkGroup>()
+                    || parent->isDerivedFrom<App::Link>()) {
+                    // We have a link chain from this document to obj, and obj is in another document -> it is an XLink target
+                    is_linked = true;
+                }
+            }
+        }
+        if (is_linked) {
+            xShapes.push_back(obj);
+            continue;
+        }
+        //not a Link and not null.  assume to be drawable.  Undrawables will be
+        // skipped later.
+        shapes.push_back(obj);
+        if (partObj) {
+            continue;
+        }
+        //don't know if this works for an XLink
+        for (auto& sub : sel.getSubNames()) {
+            if (TechDraw::DrawUtil::getGeomTypeFromName(sub) == "Face") {
+                faceName = sub;
+                //
+                partObj = obj;
+                break;
+            }
+        }
+    }
+
+    if (shapes.empty() &&
+        xShapes.empty() &&
+        !viewCreated) {
+        if (Preferences::ShowInsertFileMessage()) {
+            auto msgText = QObject::tr("If you want to insert a view from existing objects, please select them before invoking this tool. Without a selection, a file browser will open, to insert a SVG or image file.");
+            QMessageBox msgBox;
+            msgBox.setText(msgText);
+            auto dontShowMsg = QObject::tr("Do not show this message again");
+            QCheckBox dontShowCheckBox(dontShowMsg, &msgBox);
+            msgBox.setCheckBox(&dontShowCheckBox);
+            // QPushButton* okButton = msgBox.addButton(QMessageBox::Ok);
+
+            msgBox.exec();
+            if (msgBox.standardButton(msgBox.clickedButton()) == QMessageBox::Ok &&
+                dontShowCheckBox.isChecked()) {
+                Preferences::setShowInsertFileMessage(false);
+            }
+        }
+
+        QString filename = Gui::FileDialog::getOpenFileName(Gui::getMainWindow(),
+            QObject::tr("Select a SVG or Image file to open"),
+            Preferences::defaultSymbolDir(),
+            QString::fromLatin1("%1 (*.svg *.svgz *.jpg *.jpeg *.png *.bmp);;%2 (*.*)")
```

Comment:
QStringLiteral

Reply:
Yeah, I just moved the single giant function out with minimal changes.  Clean up on this one comes later.